### PR TITLE
RCD now rebuilds existing constructs faster and with less resources

### PR DESCRIPTION
**Pull request by Mothblocks**
## About The Pull Request

https://user-images.githubusercontent.com/35135081/112786932-39489f00-900c-11eb-8f92-1982137716b3.mp4

Works on grilles, windows, and walls.

I'd like to get a sound for this, so it's drafted until either I can't get one or until I get it. Code, however, is ready.

## Why It's Good For The Game

People don't rebuild rooms because it's tedious and unfun. This change makes rebuilding rooms less of a chore, while not eliminating the amount of time it takes to do (since you still will need to get hardsuits etc) and without buffing using the RCD for non-reconstruction purposes.

In the future, I want to add other features to help make reconstructing rooms less of a pain, such as showing the original wiring of the room with engineering goggles.

## Changelog
:cl:
add: The RCD now has an option to scan the world around you for showing original walls and windows.
add: The RCD will now rebuild existing windows/windows/grilles significantly faster and with less resources.
/:cl:

### CODE REVIEW

**Comment by KathyRyals** [comment]
You actually did it ahaha. Amazing. If no one steps up for the sound, I guess I could try my hand at recording one.

**Comment by Mothblocks** [comment]
i have a friend who said he'd do it but he might be busy with stuff

File: code/game/objects/items/RCD.dm [hunk 1]
```diff
@@ -1132,11 +1207,15 @@ RLD
 /obj/item/rcd_upgrade/silo_link
 	desc = "It contains direct silo connection RCD upgrade."
 	upgrade = RCD_UPGRADE_SILO_LINK
-
+_
```
Comment (Mothblocks) [inline code/game/objects/items/RCD.dm:1210]: oijvb

**Comment by Jhyrachy** [comment]
If possible, would be nice to automatically reconstruct without having to select the correct type.  
In short: enable scan -> click on highlighted tile -> automagically restored

**Comment by Mothblocks** [comment]
I think that's a good enough idea, though would have to make it so you'd have to wait until the scan is over in order to build something else there, since otherwise it would be forcing the "wrong" thing.

**Comment by Mothblocks** [comment]
i'm not sure i'll get it any time soon. i'll make a separate pr if he shoots me the sound.

File: code/game/objects/items/RCD.dm (hunk 1 above)
Comment (Mothblocks) [reply code/game/objects/items/RCD.dm:1210]:
```suggestion

```

File: code/__DEFINES/flags.dm [hunk 2]
```diff
@@ -84,6 +84,8 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
 /// If a turf cant be jaunted through.
 #define NOJAUNT (1<<0)
 #define UNUSED_RESERVATION_TURF (1<<1)
+/// This turf was initialized on mapload
+#define MAPLOAD_INTIIALIZED_TURF (1<<2)
```
Comment (AnturK) [inline code/__DEFINES/flags.dm:88]: You forgot to use this one.

File: code/game/objects/structures/grille.dm [hunk 3]
```diff
@@ -54,9 +54,20 @@
...
-				return list("mode" = RCD_WINDOWGRILLE, "delay" = 40, "cost" = 12)
-			return list("mode" = RCD_WINDOWGRILLE, "delay" = 20, "cost" = 8)
+				delay = 4 SECONDS
+				cost = 12
+
+			var/turf/grille_turf = get_turf(src)
+
+			if (grille_turf?.rcd_memory == RCD_MEMORY_WINDOWGRILLE)
+				cost /= RCD_MEMORY_COST_BUFF
+				delay /= RCD_MEMORY_SPEED_BUFF
```
Comment (AnturK) [inline code/game/objects/structures/grille.dm:68]: wrap these in helper proc to reduce copypaste maybe ?

**Review by AnturK — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/grille.dm (hunk 3 above)
Comment (Mothblocks) [reply code/game/objects/structures/grille.dm:68]: Define might be better, since procs can't return two values without a list.

Comment (Mothblocks) [reply code/game/objects/structures/grille.dm:68]: Actually, it could be a pure function of `f(x: RCDResult) -> RCDResult` that gets run, nevermind.

**Review by AnturK — APPROVED** [review]

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/__HELPERS/icons.dm [hunk 4]
```diff
@@ -943,10 +943,10 @@ world
 				I.pixel_y++
 		add_overlay(I)//And finally add the overlay.
 
-/proc/getHologramIcon(icon/A, safety=1)//If safety is on, a new icon is not created.
+/proc/getHologramIcon(icon/A, safety = TRUE, opacity = 0.5)//If safety is on, a new icon is not created.
 	var/icon/flat_icon = safety ? A : new(A)//Has to be a new icon to not constantly change the same icon.
 	flat_icon.ColorTone(rgb(125,180,225))//Let's make it bluish.
-	flat_icon.ChangeOpacity(0.5)//Make it half transparent.
+	flat_icon.ChangeOpacity(opacity)//Make it half transparent.
```
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/__HELPERS/icons.dm:949]: useless comments strike again

File: code/game/turfs/closed/walls.dm [hunk 5]
```diff
@@ -19,6 +19,8 @@
 	smoothing_groups = list(SMOOTH_GROUP_CLOSED_TURFS, SMOOTH_GROUP_WALLS)
 	canSmoothWith = list(SMOOTH_GROUP_WALLS)
 
+	rcd_memory = RCD_MEMORY_WALL
```
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/game/turfs/closed/walls.dm:22]: They *all* only work on roundstart--it's for rebuilding. Windows set the memory from the structure spawners, which are only ever created during the map process, as far as I know.

[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/turfs/closed/walls.dm:22]: ah I see, thanks

**Comment by AnturK** [comment]
Would be nice to squash this and blueprints together later.

**Pull request merged**